File: 129-sum-root-to-leaf-numbers/129-sum-root-to-leaf-numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        global res, final
        res = 0
        final = 0
        def dfs(root):
            global res, final
            if not root.left and not root.right:
                final += res * 10 + root.val
                return
            res = res * 10 + root.val
            if root.left:
                dfs(root.left)
            if root.right:
                dfs(root.right)
            res = (res - root.val) // 10
        dfs(root)
        return final

File: 129-sum-root-to-leaf-numbers/129-sum-root-to-leaf-numbers.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int res = 0;
    int sum = 0;
    public int sumNumbers(TreeNode root) {
        //@input : root of a tree
        //@output : the sum of all root-to-leaf numbers
        //edge case : null
        //breaking down:
        //1.pre-order traversal to find the corresponding number for each root
        //when a leaf is reached, the result is added into the sum
        if(root.left == null && root.right == null){
            sum = sum * 10 + root.val;
            res += sum;
            sum = (sum - root.val) / 10;
            return res;
        }
        sum = sum * 10 + root.val;
        if(root.left != null){
            sumNumbers(root.left);
        }
        if(root.right != null){
            sumNumbers(root.right);
        }
        sum = (sum - root.val) / 10;
        return res;
    }
}